README.md
# Color

Extremely tiny color package (1300 bytes gzipped) for parsing, manipulating, and computing CSS colors.

In particular, supports most of the algorithms needed for WCAG accessibility guidelines, and a
swatch algorithm superior to manipulating "lightness."


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

### 1.0.1 (2021-05-30)


test/index.test.ts
import 'jest';      // Ref: https://jestjs.io/docs/en/expect#reference
import Color from "../src";


test("construction", () => {
  expect(new Color(0, 0, 0, 0).toCss()).toBe('#00000000');
  expect(new Color(1, 1, 1, 1).toCss()).toBe('#ffffffff');
  expect(new Color(0.5, 0.75, 0.25, 0.0625).toCss()).toBe('#80bf4010');
  expect(new Color(0.502, 0.753, 0.25, 0.0625).toCss()).toBe('#80c04010');
  expect(new Color(-1, -0.001, 1.0001, 2).toCss()).toBe('#0000ffff');
});

test("invalid color", () => {
  function t(p: any) { expect(Color.fromCss(p).toCss()).toBe('#00000000') }

  t('')
  t('tacos')
  t('x')
  t(123)
  t(null)
  t(undefined)
  t({ r: 123, g: 123, b: 123 })
});

test("named colors", () => {
  expect(Color.fromCss("transparent").toCss()).toBe('#00000000')
  expect(Color.fromCss("black").toCss()).toBe('#000000ff')
  expect(Color.fromCss("white").toCss()).toBe('#ffffffff')
});

test("from #RGB(A)", () => {
  function t(p: string, q: string) { expect(Color.fromCss(p).toCss()).toBe(q) }

  // full string
  t('#0000', '#00000000')
  t('#ffff', '#ffffffff')
  t('#1234', '#11223344')
  t('#edbc', '#eeddbbcc')

  // without alpha channel
  t('#000', '#000000ff')
  t('#fff', '#ffffffff')
  t('#123', '#112233ff')
  t('#edb', '#eeddbbff')
});

test("from #RRGGBB(AA)", () => {
  function t(p: string, q: string) { expect(Color.fromCss(p).toCss()).toBe(q) }

  // full string
  t('#00000000', '#00000000')
  t('#ffffffff', '#ffffffff')
  t('#80c04010', '#80c04010')
  t('#fe01c912', '#fe01c912')
  t('#0000ffff', '#0000ffff')

  // without alpha channel
  t('#000000', '#000000ff')
  t('#ffffff', '#ffffffff')
  t('#80c040', '#80c040ff')
  t('#fe01c9', '#fe01c9ff')
  t('#0000ff', '#0000ffff')
});

test("from rgb(r,g,b(,a))", () => {
  function t(r: number, g: number, b: number, a: number) {
    const target = new Color(r / 255, g / 255, b / 255, a / 255).toCss();
    const targetNoAlpha = new Color(r / 255, g / 255, b / 255, 1).toCss();

    expect(Color.fromCss(`rgb(${r},${g},${b},${a})`).toCss()).toBe(target)
    expect(Color.fromCss(`rgbA(${r},${g},${b},${a})`).toCss()).toBe(target)
    expect(Color.fromCss(`RGB(${r},${g},${b})`).toCss()).toBe(targetNoAlpha)
    expect(Color.fromCss(`rgb(${r * 100 / 255}%,${g * 100 / 255}%,${b * 100 / 255}%,${a * 100 / 255}%)`).toCss()).toBe(target)
    expect(Color.fromCss(`rgb(${r * 100 / 255}%,${g * 100 / 255}%,${b * 100 / 255}%)`).toCss()).toBe(targetNoAlpha)
  }

  t(0, 0, 0, 0)
  t(1, 2, 3, 4)
  t(255, 255, 255, 255)
  t(255, 254, 253, 252)
  t(179, 180, 181, 182)
});

test("from hsv(h,s,v(,a))", () => {
  function t(r: number, g: number, b: number, h: number, s: number, v: number) {
    const target = new Color(r / 255, g / 255, b / 255, 1).toCss();

    expect(Color.fromCss(`hsv(${h},${s},${v})`).toCss()).toBe(target)
    expect(Color.fromCss(`HSVA(${h},${s},${v},100%)`).toCss()).toBe(target)
  }

  t(0, 0, 0, 0, 0, 0)
  t(255, 255, 255, 0, 0, 100)
  t(255, 0, 0, 0, 100, 100)
  t(0, 255, 0, 120, 100, 100)
  t(0, 0, 255, 240, 100, 100)
  t(255, 255, 0, 60, 100, 100)
  t(0, 255, 255, 180, 100, 100)
  t(255, 0, 255, 300, 100, 100)
  t(191, 191, 191, 0, 0, 75)
  t(128, 128, 128, 0, 0, 50)
  t(128, 0, 0, 0, 100, 50)
  t(128, 128, 0, 60, 100, 50)
  t(0, 128, 0, 120, 100, 50)
  t(0, 0, 128, 240, 100, 50)
  t(18, 18, 18, 0, 0, 7.1)
});

test("from hsl(h,s,l(,a))", () => {
  function t(r: number, g: number, b: number, h: number, s: number, l: number) {
    const target = new Color(r / 255, g / 255, b / 255, 1).toCss();

    expect(Color.fromCss(`hsl(${h},${s},${l})`).toCss()).toBe(target)
    expect(Color.fromCss(`HSLA(${h},${s},${l},100%)`).toCss()).toBe(target)
  }

  t(0, 0, 0, 0, 0, 0)
  t(255, 255, 255, 0, 0, 100)
  t(255, 0, 0, 0, 100, 50)
  t(0, 255, 0, 120, 100, 50)
  t(0, 0, 255, 240, 100, 50)
  t(255, 255, 0, 60, 100, 50)
  t(0, 255, 255, 180, 100, 50)
  t(255, 0, 255, 300, 100, 50)
  t(191, 191, 191, 0, 0, 75)
  t(128, 128, 128, 0, 0, 50)
  t(128, 0, 0, 0, 100, 25)
  t(127, 128, 0, 60, 100, 25)   // Really 128 for red?  Or is that debatable anyway?
  t(0, 128, 0, 120, 100, 25)
  t(0, 0, 128, 240, 100, 25)
  t(18, 18, 18, 0, 0, 7.1)
});

test("brightness", () => {
  function t(css: string, b: number) {
    const c = Color.fromCss(css);
    expect(c.getBrightness()).toBeCloseTo(b, 2)
    expect(c.isLight()).toBe(c.getBrightness() >= 0.5)
  }

  t('#fff', 1)
  t('#000', 0)
  t('#7f7f7f', 0.5)
  t('#ff0000', 0.299)
  t('#00ff00', 0.587)
  t('#0000ff', 0.114)
});

test("luminance", () => {
  function t(c: string, b: number) {
    expect(Color.fromCss(c).getLuminance()).toBeCloseTo(b, 2)
  }

  t('#fff', 1)
  t('#000', 0)
  t('#7f7f7f', 0.2122)
  t('#ff0000', 0.2126)
  t('#00ff00', 0.7152)
  t('#0000ff', 0.0722)
});

test("contrast ratio", () => {
  function t(css1: string, css2: string, x: number) {
    const c1 = Color.fromCss(css1);
    const c2 = Color.fromCss(css2);
    expect(c1.getContrastRatio(c2)).toBeCloseTo(x, 2)
    expect(c2.getContrastRatio(c1)).toBeCloseTo(x, 2)
  }

  t('#fff', '#fff', 1);
  t('#000', '#000', 1);
  t('#fff', '#000', 21);
  t('#121212', '#000000', 1.12);
  t('#121212', '#ffffff', 18.73);
  t('#ff0000', '#0000ff', 2.15);
  t('#ff8080', '#0000ff', 3.54);
});

test("mix", () => {
  function t(css1: string, css2: string, p: number, cssResult: string) {
    const c1 = Color.fromCss(css1);
    const c2 = Color.fromCss(css2);
    expect(c1.mix(c2, p).toCss()).toBe(cssResult)
  }

  for (const css of ['#000000ff', '#ffffff00', '#c3512f45']) {
    t(css, css, 1.2, css);
    t(css, css, 1, css);
    t(css, css, 0.5, css);
    t(css, css, 0, css);
    t(css, css, -0.2, css);
  }
  t('#fff', '#000', 0, '#ffffffff')
  t('#fff', '#000', 1, '#000000ff')
  t('#fff', '#000', 0.5, '#808080ff')
  t('#fff', '#000', 0.75, '#404040ff')
  t('#fff', '#000', 0.25, '#bfbfbfff')
  t('#c0804020', '#4020ff40', 1, '#4020ff40')
  t('#c0804020', '#4020ff40', 0, '#c0804020')
  t('#c0804020', '#4020ff40', 0.5, '#8050a030')
  t('#c0804020', '#4020ff40', 0.75, '#6038cf38')
});

test("brighten", () => {
  function t(css: string, p: number, cssResult: string) {
    expect(Color.fromCss(css).brighten(p).toCss()).toBe(cssResult)
  }

  t('#c0804020', 0, '#c0804020')
  t('#c0804020', 1, '#ffffff20')
  t('#c0804020', 0.5, '#e0c0a020')
});

test("darken", () => {
  function t(css: string, p: number, cssResult: string) {
    expect(Color.fromCss(css).darken(p).toCss()).toBe(cssResult)
  }

  t('#c0804020', 0, '#c0804020')
  t('#c0804020', 1, '#00000020')
  t('#c0804020', 0.5, '#60402020')
});

test("distance", () => {
  function t(css1: string, css2: string, x: number) {
    const c1 = Color.fromCss(css1);
    const c2 = Color.fromCss(css2);
    expect(c1.getPerceptualDistance(c2)).toBeCloseTo(x, 2)
    expect(c2.getPerceptualDistance(c1)).toBeCloseTo(x, 2)
  }

  t('#000', '#000', 0)
  t('#fff', '#fff', 0)
  t('#000', '#fff', 1)
  t('#808080', '#fff', 0.5)
  t('#808080', '#000', 0.5)
  t('#ff0000', '#000', 0.55)
  t('#00ff00', '#000', 0.77)
  t('#0000ff', '#000', 0.34)
  t('#ff0000', '#0000ff', 0.64)
  t('#ff0000', '#00ff00', 0.94)
});


src/index.ts

// Global regular expressions
const RE_THREE_DIGIT = /#([0-9a-fA-F])([0-9a-fA-F])([0-9a-fA-F])([0-9a-fA-F])?\b/i;
const RE_SIX_DIGIT = /#([0-9a-fA-F]{2})([0-9a-fA-F]{2})([0-9a-fA-F]{2})([0-9a-fA-F]{2})?\b/i;
const RE_LONGFORM = /\b(rgb|hs[vl])a?\s*\(\s*([\.\d]+\%?)\s*,\s*([\.\d]+\%?)\s*,\s*([\.\d]+\%?)\s*(?:,\s*([\.\d]+\%?))?\s*\)/i;

// Global functions for speed
const round = Math.round
const pow = Math.pow
const min = Math.min
const max = Math.max
const sqrt = Math.sqrt

// Returns the squared difference between two numbers
function sqdiff(a: number, b: number): number {
  const d = a - b;
  return d * d;
}

// Parse string hex digits into a number
function fromHex(s: string): number {
  return parseInt(s, 16);
}

// Parse a number, converting from `[0,maximum]` to `[0,1]`, or as a percentage if that is the trailing character
function fromNumber(s: string, maximum: number): number {

  // Compute an initial [0,1] value
  let x;
  if (s.endsWith('%')) x = parseFloat(s) / 100;
  else x = parseFloat(s) / maximum;

  // Clamp and slight round-off at the extremities, which are important values for color algorithms
  return x < 0.00001 ? 0 : x > 0.99999 ? 1 : x;
}

/**
 * Converts a number in `[0,1]` into a 2-digit hex string in the range `[0,255]`.  Clamps if outside that range.
 */
function h2(x: number): string {
  x = round(x * 255)
  if (x < 0) return '00'
  if (x < 16) return '0' + x.toString(16)
  if (x < 255) return x.toString(16)
  return 'ff'
}

export default class Color {

  /**
   * Create a new, mutable color object, given color channels in the range `[0,1]`.
   * Typically it's easier to use a static method to create a color.
   */
  constructor(public r: number, public g: number, public b: number, public a: number) {
    // nothing else to do
  }

  /**
   * Reads a color from a CSS string, or returns `null` if the color is invalid.
   * Supports 'transparent' but no other written colors.
   */
  static fromCss(s: string | null | undefined): Color {
    if (typeof s !== 'string') return NAMED_COLORS.transparent.clone();
    let m: RegExpMatchArray | null;
    const c = NAMED_COLORS[s];
    if (c) return c.clone();
    if (m = s.match(RE_SIX_DIGIT)) {
      return new Color(fromHex(m[1]) / 255, fromHex(m[2]) / 255, fromHex(m[3]) / 255, m[4] ? fromHex(m[4]) / 255 : 1)
    }
    if (m = s.match(RE_THREE_DIGIT)) {
      return new Color(fromHex(m[1] + m[1]) / 255, fromHex(m[2] + m[2]) / 255, fromHex(m[3] + m[3]) / 255, m[4] ? fromHex(m[4] + m[4]) / 255 : 1)
    }
    if (m = s.match(RE_LONGFORM)) {
      const a = m[5] ? fromNumber(m[5], 255) : 1;
      switch (m[1].toLowerCase()) {
        case 'rgb': return new Color(fromNumber(m[2], 255), fromNumber(m[3], 255), fromNumber(m[4], 255), a);
        case 'hsv': return Color.fromHSV(fromNumber(m[2], 360), fromNumber(m[3], 100), fromNumber(m[4], 100), a);
        case 'hsl': return Color.fromHSL(fromNumber(m[2], 360), fromNumber(m[3], 100), fromNumber(m[4], 100), a);
        /* istanbul ignore next */
        default: return new Color(0, 0, 0, 0);    // unknown format, but regex largely prevents this from happening!
      }
    }
    return NAMED_COLORS.transparent.clone();
  }

  /**
   * Converting from HSV color space, all numbers in `[0,1]`.
   */
  static fromHSV(h: number, s: number, v: number, a: number): Color {
    h = h * 6;
    const i = Math.floor(h),
      f = h - i,
      p = v * (1 - s),
      q = v * (1 - f * s),
      t = v * (1 - (1 - f) * s),
      mod = i % 6,
      r = [v, q, p, p, t, v][mod],
      g = [t, v, v, q, p, p][mod],
      b = [p, p, t, v, v, q][mod];

    return new Color(r, g, b, a);
  }

  /**
   * Converting from HSL color space, all numbers in `[0,1]`.
   */
  static fromHSL(h: number, s: number, l: number, a: number): Color {
    var r, g, b;

    function hue2rgb(p: number, q: number, t: number) {
      if (t < 0) t += 1;
      if (t > 1) t -= 1;
      if (t < 1 / 6) return p + (q - p) * 6 * t;
      if (t < 1 / 2) return q;
      if (t < 2 / 3) return p + (q - p) * (2 / 3 - t) * 6;
      return p;
    }

    if (s === 0) {
      r = g = b = l; // achromatic
    }
    else {
      var q = l < 0.5 ? l * (1 + s) : l + s - l * s;
      var p = 2 * l - q;
      r = hue2rgb(p, q, h + 1 / 3);
      g = hue2rgb(p, q, h);
      b = hue2rgb(p, q, h - 1 / 3);
    }

    return new Color(r, g, b, a);
  }

  /**
   * Returns a copy of this color object.  Useful because they are mutable.
   */
  clone(): Color {
    return new Color(this.r, this.g, this.b, this.a);
  }

  /**
   * Returns a CSS color value representation, e.g. `#f03452ff`.
   */
  toCss(): string {
    return '#' + h2(this.r) + h2(this.g) + h2(this.b) + h2(this.a)
  }

  /**
   * Gets the perceived brightness of the color, in `[0,1]`, as defined by
   * http://www.w3.org/TR/AERT#color-contrast
   */
  getBrightness(): number {
    return this.r * 0.299 + this.g * 0.587 + this.b * 0.114;
  }

  /**
   * Gets the distance between this color and another as `[0,1]` where `0` is identical and `1` is the distance from white to black,
   * weighted by human's perception of color differences.
   */
  getPerceptualDistance(c: Color): number {
    return sqrt(sqdiff(this.r, c.r) * 0.299 + sqdiff(this.g, c.g) * 0.587 + sqdiff(this.b, c.b) * 0.114);
  }

  /**
   * Gets the luminance of the color, in `[0,1]`, as defined by
   * http://www.w3.org/TR/2008/REC-WCAG20-20081211/#relativeluminancedef
   */
  getLuminance(): number {
    const RsRGB = this.r, GsRGB = this.g, BsRGB = this.b;
    var R, G, B;

    if (RsRGB <= 0.03928) { R = RsRGB / 12.92; } else { R = pow(((RsRGB + 0.055) / 1.055), 2.4); }
    if (GsRGB <= 0.03928) { G = GsRGB / 12.92; } else { G = pow(((GsRGB + 0.055) / 1.055), 2.4); }
    if (BsRGB <= 0.03928) { B = BsRGB / 12.92; } else { B = pow(((BsRGB + 0.055) / 1.055), 2.4); }
    return (0.2126 * R) + (0.7152 * G) + (0.0722 * B);
  }

  /**
   * True if this color's brightness is above `0.5`.  Generally "light" background colors should receive
   * dark foreground (e.g. text) colors, and vice versa.
   */
  isLight(): boolean {
    return this.getBrightness() >= 0.5;
  }

  /**
   * Computes the Contrast Ratio (i.e. readability) between this color and another, as defined by
   * http://www.w3.org/TR/2008/REC-WCAG20-20081211/#contrast-ratiodef
   */
  getContrastRatio(c: Color): number {
    const l1 = this.getLuminance();
    const l2 = c.getLuminance();
    return (max(l1, l2) + 0.05) / (min(l1, l2) + 0.05);
  }

  /**
   * Mixes another color into this color and return in-place.  The resulting color is `p` the new color,
   * and `1-p` the existing color, with `p` in `[0,1]`.
   */
  mix(c: Color, p: number): this {
    if (p <= 0) return this;   // not mixing any of p, actually
    if (p >= 1) p = 1;    // totally the other color
    this.r += (c.r - this.r) * p;
    this.g += (c.g - this.g) * p;
    this.b += (c.b - this.b) * p;
    this.a += (c.a - this.a) * p;
    return this;
  }

  /**
   * Brighten and return in-place, by an amount in `[0,1]`, where `0` does not change the color,
   * and `1` makes the color completely white.  Does not alter the opacity.
   */
  brighten(x: number): this {
    return this.mix(new Color(1, 1, 1, this.a), x);
  }

  /**
   * Darken and return in-place, by an amount in `[0,1]`, where `0` does not change the color,
   * and `1` makes the color completely black.  Does not alter the opacity.
   */
  darken(x: number): this {
    return this.mix(new Color(0, 0, 0, this.a), x);
  }

}

// Color name table -- just some basics, not the full set of HTML colors that no one uses
const NAMED_COLORS: Record<string, Color> = {
  "transparent": new Color(0, 0, 0, 0),
  "white": new Color(1, 1, 1, 1),
  "black": new Color(0, 0, 0, 1),
}
